File: algos/buyAndSellStock/buyAndSellStock.go
func maxProfit(prices []int) int {
	greatestProfitSoFar := 0
	lowestNumSeenSoFar := prices[0]

	for _, price := range prices {
		if (price - lowestNumSeenSoFar) > greatestProfitSoFar {
			greatestProfitSoFar = price - lowestNumSeenSoFar
		}
		if price < lowestNumSeenSoFar {
			lowestNumSeenSoFar = price
		}
	}

	return greatestProfitSoFar
}

File: algos/buyAndSellStock/buyAndSellStock.ts
var maxProfit = function (prices) {
  let lowestNumSeenSoFar = prices[0];
  let maxProfitMadeSoFar = 0;

  for (let i = 0; i < prices.length; i++) {
    const current = prices[i];
    const profit = current - lowestNumSeenSoFar;
    if (profit > maxProfitMadeSoFar) {
      maxProfitMadeSoFar = profit;
    }
    if (current < lowestNumSeenSoFar) {
      lowestNumSeenSoFar = current;
    }
  }

  return maxProfitMadeSoFar;
};